core/src/main/java/org/apache/druid/data/input/impl/RetryingInputStream.java

```diff
@@ -131,76 +136,92 @@ private static void throwAsIOException(Throwable t) throws IOException
   @Override
   public int read() throws IOException
   {
-    for (int nTry = 0; nTry < maxRetry; nTry++) {
+    openIfNeeded();
+
+    for (int nTry = 0; nTry < maxTries; nTry++) {
       try {
         return delegate.read();
       }
       catch (Throwable t) {
         waitOrThrow(t, nTry);
       }
     }
-    return delegate.read();
+
+    // Can't happen, because the final waitOrThrow would have thrown.
+    throw new IllegalStateException();
   }
 
   @Override
   public int read(byte[] b) throws IOException
   {
-    for (int nTry = 0; nTry < maxRetry; nTry++) {
-      try {
-        return delegate.read(b);
-      }
-      catch (Throwable t) {
-        waitOrThrow(t, nTry);
-      }
-    }
-    return delegate.read(b);
+    return read(b, 0, b.length);
```

Comment:
I overrode this method just in case `delegate.read(byte[])` does something else than simply calling `read(b, 0, b.length)` even though it's very rare (and likely a wrong implementation). I think it would be safer to keep it.

Reply:
Sure, I'll restore it.